proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,319 @@
+
+### MainActor
+
+The `MainActor` provides a static rethrowing `run` method. This method can also
+adopt typed throws:
+
+```swift
+public static func run<T: Sendable, Failure: Error>(
+  resultType: T.Type = T.self,
+  body: @MainActor @Sendable () throws(Failure) -> T
+) async rethrows -> T
+```
+
+### `AsyncSequence` & `AsyncIteratorProtocol`
+
+`AsyncSequence` iterators can throw during iteration, as described by the
+`throws` on the `next()` operation on async iterators:
+
+```swift
+public protocol AsyncIteratorProtocol {
+  associatedtype Element
+  mutating func next() async throws -> Element?
+}
+```
+
+Introduce a new associated type `Failure` into this protocol to use as the
+thrown error type of `next()`, i.e.,
+
+```swift
+associatedtype Failure: Error = any Error
+mutating func next() async throws(Failure) -> Element?
+```
+
+Then introduce an associated type `Failure` into `AsyncSequence` that provides a
+more convenient name for this type, i.e.,
+
+```swift
+associatedtype Failure where AsyncIterator.Failure == Failure
+```
+
+With the new `Failure` associated type, async sequences can be composed without
+losing information about whether (and what kind) of errors they throw.
+
+With the new `Failure` type in place, we can adopt [primary asociated
+types](https://github.com/apple/swift-evolution/blob/main/proposals/0346-light-weight-same-type-syntax.md)
+for these protocols:
+
+```swift
+public protocol AsyncIteratorProtocol<Element, Failure> {
+  associatedtype Element
+  associatedtype Failure: Error = any Error
+  mutating func next() async throws(Failure) -> Element?
+}
+
+public protocol AsyncSequence<Element, Failure> {
+  associatedtype AsyncIterator: AsyncIteratorProtocol
+  associatedtype Element where AsyncIterator.Element == Element
+  associatedtype Failure where AsyncIterator.Failure == Failure
+  __consuming func makeAsyncIterator() -> AsyncIterator
+}
+```
+
+This allows the use of `AsyncSequence` with both opaque types (`some
+AsyncSequence<String, any Error>`) and existential types (`any
+AsyncSequence<Image, NetworkError>`). 
+
+### AsyncThrowingStream
+
+The non-throwing variant `AsyncStream` cannot adopt typed throws
+since it does not have a generic `Failure` parameter; however,
+the throwing variant `AsyncThrowingStream` already contains a `Failure`
+generic parameter and can be extended to adopt typed throws.
+The current initializers of `AsyncThrowingStream` are constraining
+the `Failure` type to `Error`.
+
+```swift
+public init(
+  _ elementType: Element.Type = Element.self,
+  bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded,
+  _ build: (Continuation) -> Void
+) where Failure == Error
+
+public init(
+  unfolding produce: @escaping () async throws -> Element?
+) where Failure == Error 
+```
+
+We propose to add two new initializers that drop the `Failure` constraint:
+
+```swift
+public init(
+  _ elementType: Element.Type = Element.self,
+  _ failureType: Failure.Type = Failure.self,
```

Comment:
This init seems inconsistent, we usually do the `Group(of: Intl.self` / `Stream(of: Int.self)`, should this be

```suggestion
init(
  of elementType: Element.Type = Element.self,
  failureType: Failure.Type = Failure.self,
...)
```

Reply:
Well the current APIs are not following that, but I will align it.

Reply:
Yeah, I know they;'re not following the pattern and it's very annoying that it's visible in API "which person did which API" -- time to fix it! 😄